src/software/ai/evaluation/possession.cpp

```diff
@@ -2,10 +2,125 @@
 
 #include "shared/constants.h"
 #include "software/ai/evaluation/intercept.h"
+#include "software/geom/algorithms/distance.h"
 #include "software/world/ball.h"
 #include "software/world/field.h"
 #include "software/world/team.h"
 
+PossessionTracker::PossessionTracker() {}
+
+TeamPossession PossessionTracker::getTeamWithPossession(const Team &friendly_team,
+                                                        const Team &enemy_team,
+                                                        const Ball &ball,
+                                                        const Field &field)
+{
+    // Based on CMDragons TDP 2015
+
+    updateTimes(friendly_team, enemy_team, ball, field);
+
+    const Duration TIME_NEAR_THRESHOLD = Duration::fromSeconds(TIME_NEAR_THRESHOLD_S);
+    const Duration TIME_FAR_THRESHOLD  = Duration::fromSeconds(TIME_FAR_THRESHOLD_S);
+
+    if ((time_near_friendly > TIME_NEAR_THRESHOLD) &&
+        (time_near_enemy < TIME_NEAR_THRESHOLD))
+    {
+        possession = TeamPossession::FRIENDLY_TEAM;
+    }
+    else if ((time_near_friendly < TIME_NEAR_THRESHOLD) &&
+             (time_near_enemy > TIME_NEAR_THRESHOLD))
+    {
+        possession = TeamPossession::ENEMY_TEAM;
+    }
+    else if ((time_near_friendly > TIME_NEAR_THRESHOLD) &&
+             (time_near_enemy > TIME_NEAR_THRESHOLD))
+    {
+        // Both teams are considered to have presence over the ball.
+        // If the ball is on our side of the field, or no enemy robots are
+        // on our side of the field, consider enemy team as having possession.
```

Comment:
Good idea re: breakbeam. For some scenarios the breakbeam data should be useful. E.g. If the breakbeam is triggered and we know that the enemy team is far away from ball, then we can probably be confident that we have possession.

Maybe we should introduce some different types of possession that trigger more cautious strategies in OffensePlay. E.g. in the scenario you described, we could return something like `TeamPossession::FRIENDLY_GUARDING` to indicate that possession is being contested and that the friendly team should try to protect the ball. This could initiate a more offensive defense with a ball retrieval/dribble away tactic

Reply:
That's a great idea! It'd be really interesting having different levels of offense/defense depending on the scenario and the placement of the robots. 

> E.g. If the breakbeam is triggered and we know that the enemy team is far away from ball, then we can probably be confident that we have possession.

That's right, though we do have to be slightly careful, since there could be scenarios where we have the ball in our dribbler, but the enemy robot is right in our face (potentially with the ball touching their dribbler as well) trying to steel the ball from us. Switching to an offensive play in that scenario is risky.